lib/tessel/deploy.js

```diff
@@ -383,10 +397,168 @@ actions.glob = {
   }
 };
 
+actions.resolveBinaryModules = function(opts) {
+  var cwd = process.cwd();
+  var target = opts.target || cwd;
+  var relative = path.relative(cwd, target);
+  var globRoot = relative || target;
+
+  return new Promise(function(resolve, reject) {
+    // Find all modules that include a compiled binary
+    var pattern = 'node_modules/**/*.node';
+    var binaries = actions.glob.files(globRoot, [pattern]).map(function(mPath) {
+      // Garher information about each found module
+      var binName = path.basename(mPath);
+      var modulePath = binaryModulePath(path.dirname(mPath));
+      var packageJson = require(path.join(globRoot, modulePath, 'package.json'));
+      return {
+        binName: binName,
+        name: packageJson.name,
+        version: packageJson.version,
+        path: modulePath,
+        packageJson: packageJson,
+      };
+    });
+
+    // Using the discovered binaries, create a Promise to
+    // represent the request, receipt, and extraction process
+    // for each of them.
+    var requests = binaries.map(function(details) {
+      return new Promise(function(resolve, reject) {
+        var tgz = `${details.name}-${details.version}-Release.tgz`;
+
+        // Store the name of the path where this might already
+        // be cached, but will most certainly be cached once
+        // it has been resolved.
+        details.extractPath = path.join(BINARY_CACHE_PATH, path.basename(tgz, '.tgz'));
+
+        // Sometimes a module will depend on other
+        // modules that in turn depend on the same modules
+        // as eachother. Since we only need one copy of a given
+        // module, resolve this entry without any further action.
+        if (binaryModulesUsed.has(details.binName)) {
+          return resolve();
+        }
+
+        // Update running list of binary modules that
+        // this project is using.
+        binaryModulesUsed.set(details.binName, details);
+
+        // If an extraction path already exists locally,
+        // resolve this entry without any further action.
+        if (fs.existsSync(details.extractPath)) {
+          return resolve();
+        } else {
+          //
+          // Request, receive, unzip, extract and resolve
+          //
+          var url = urljoin(BINARY_SERVER_ROOT, tgz);
+
+          // Make a ~/.tessel/binaries/MODULE-NAME directory
+          fs.mkdirp(details.extractPath, function() {
+            // wget the tgz, save as
+            // ~/.tessel/binaries/MODULE-NAME.tgz
+            var gunzip = zlib.createGunzip();
+            var extract = tar.Extract({
+              path: details.extractPath,
+            });
+
+            request({
```

Comment:
@rwaldron we may want to print out a special error message if the user doesn't have internet connection or they can't reach the server for whatever reason. Most people may not even know that we're fetching these modules from the internet. I was thinking something to the extent of:
``` Unable to fetch remote binary dependencies required by [BINARY MODULE NAMES]. Please ensure you are connected to the internet before attempting to deploy this project.```